### ISSUE TITLE
[tune] PAUSED trials still occupy GPUs

### ISSUE THREAD

**Issue by ch1ck** [ray-project/ray#17720]
### What is the problem?
I am trying to customize my own scheduler. 
In this process, I want to pause those unpromising trials and release GPUs to start new trials. 
But I found that neither `return TrialScheduler.PAUSE` in on_trial_result nor using `trial_runner.trial_executor.pause_trial(trial)` can release GPUs resources. 

I printed the status of trials and used `ray.cluster_resources().get("GPU", 0)`/ `ray.available_resources().get("GPU", 0))` to get the the number of total / available GPUs,
the result is like:
082e0_00000 TERMINATED
082e0_00001 TERMINATED
082e0_00002 TERMINATED
082e0_00003 TERMINATED
082e0_00004 PAUSED
082e0_00005 PAUSED
=======================Sys info==============================
Total GPUs:  4.0
Available GPUs:  2.0

It seems that paused trials still occupy GPU resources.

But I saw the annotation in trial_executor.py:  
```
  def pause_trial(self, trial):
      """Pauses the trial.
      We want to release resources (specifically GPUs) when pausing an
      experiment. This results in PAUSED state that similar to TERMINATED.
      """
```

### Reproduction (REQUIRED)
```
def __init__(self):
    self.count = 0

def on_trial_result(self, trial_runner: "trial_runner.TrialRunner",
                    trial: Trial, result: Dict) -> str:
    if(result['training_iteration'] > 5 and self.count < 4):
        self.count += 1
        return TrialScheduler.STOP
    
    elif(result['training_iteration'] > 5 and self.count >= 4):
        # trial_runner.trial_executor.pause_trial(trial)
        self.count += 1
        return TrialScheduler.PAUSE

    return TrialScheduler.CONTINUE
```

in tune.run()
```
    lr_list = list()
    momentum_list = list()
    random.seed(1)
    for i in range(0, 2):
        lr_list.append(random.uniform(0.001, 0.1))

    for i in range(0, 3):
        momentum_list.append(random.uniform(0.1, 0.9))

    analysis = tune.run(
        TrainMNIST,
        metric="mean_accuracy",
        mode="max",
        scheduler=test_scheduler,
        stop={
            # "mean_accuracy": 0.99,
            "training_iteration": 3 if args.smoke_test else 20,
        },
        resources_per_trial={
            "cpu": 1,
            "gpu": 1
        },
        num_samples=1 if args.smoke_test else 1,
        checkpoint_at_end=True,
        checkpoint_freq=3,
        config={
            "args": args,
            "lr": tune.grid_search(lr_list),
            "momentum": tune.grid_search(momentum_list),
        })

    print("Best config is:", analysis.get_best_config(metric="mean_accuracy", mode="max"))
```

**Comment by richardliaw** [comment]
In `TrainMNIST` maybe you can also use `tune.utils.wait_for_gpu`: https://docs.ray.io/en/master/tune/api_docs/trainable.html?highlight=wait_for_gpu#ray.tune.utils.wait_for_gpu

**Comment by ch1ck** [comment]
> In `TrainMNIST` maybe you can also use `tune.utils.wait_for_gpu`: https://docs.ray.io/en/master/tune/api_docs/trainable.html?highlight=wait_for_gpu#ray.tune.utils.wait_for_gpu

Hi, I added `tune.utils.wait_for_gpu` to TrainMNIST step() function but it didn't change the result.

Actually I tried to pause more trials, like 
```
def on_trial_result(self, trial_runner: "trial_runner.TrialRunner",
                    trial: Trial, result: Dict) -> str:
    if(result['training_iteration'] > 5 and self.count < 1):
        self.count += 1
        return TrialScheduler.STOP
    
    elif(result['training_iteration'] > 5 and self.count >= 1):
        # trial_runner.trial_executor.pause_trial(trial)
        self.count += 1
        return TrialScheduler.PAUSE

    return TrialScheduler.CONTINUE
```

the result was

```
==============================================================
9fe49_00000 PAUSED
9fe49_00001 PAUSED
9fe49_00002 PAUSED
9fe49_00003 TERMINATED
9fe49_00004 PAUSED
9fe49_00005 PAUSED
=======================Sys info==============================
Total GPUs:  4.0
Available GPUs:  0
```
that It even has no available GPUs to run any trials.

**How can I correct my scheduler to make paused trials release their GPUs resources?**
Thanks

**Comment by richardliaw** [comment]
Can you post your full script? I'm not quite sure why you're seeing this behavior. Pausing should release all resources of a given trial.

**Comment by ch1ck** [comment]
> Can you post your full script? I'm not quite sure why you're seeing this behavior. Pausing should release all resources of a given trial.

Yes, the expectation is that the available GPUs should be equal to total GPUs when all trials are in PAUSED or TERMINATED status.

test_scheduler.py
```
from typing import Dict, Optional

from ray.tune import trial_runner
from ray.tune.result import DEFAULT_METRIC
from ray.tune.trial import Trial
from ray.tune.schedulers.trial_scheduler import TrialScheduler
import ray

class TestScheduler(TrialScheduler):
    """Simple scheduler that just runs trials in submission order."""

    def __init__(self):
        self.count = 0
        self.total_gpus = ray.cluster_resources().get("GPU", 0)

    def print_info(self, position):
        print("=======================Sys info==============================")
        print("Total GPUs: ", self.total_gpus)
        print("Available GPUs: ", ray.available_resources().get("GPU", 0))
        print(position)
        print("==============================================================")

    def on_trial_add(self, trial_runner: "trial_runner.TrialRunner",
                     trial: Trial):
        pass

    def on_trial_error(self, trial_runner: "trial_runner.TrialRunner",
                       trial: Trial):
        pass

    def on_trial_result(self, trial_runner: "trial_runner.TrialRunner",
                        trial: Trial, result: Dict) -> str:
        if(result['training_iteration'] > 5 and self.count < 1):
            self.count += 1
            return TrialScheduler.STOP
        
        elif(result['training_iteration'] > 5 and self.count >= 1):
            # trial_runner.trial_executor.pause_trial(trial)
            self.count += 1
            return TrialScheduler.PAUSE

        return TrialScheduler.CONTINUE

    def on_trial_complete(self, trial_runner: "trial_runner.TrialRunner",
                          trial: Trial, result: Dict):
        pass

    def on_trial_remove(self, trial_runner: "trial_runner.TrialRunner",
                        trial: Trial):
        pass

    def choose_trial_to_run(
            self, trial_runner: "trial_runner.TrialRunner") -> Optional[Trial]:

        self.print_info("CTTR")

        for trial in trial_runner.get_trials():
            print(trial.trial_id, trial.status)
            
        for trial in trial_runner.get_trials():
            if (trial.status == Trial.PENDING
                    and trial_runner.has_resources_for_trial(trial)):
                return trial
        # for trial in trial_runner.get_trials():
        #     if (trial.status == Trial.PAUSED
        #             and trial_runner.has_resources_for_trial(trial)):
        #         return trial
        return None

    def debug_string(self) -> str:
        return "Using FIFO scheduling algorithm."
```

test.py
```
from __future__ import print_function

import argparse
import os
import random
import torch
import torch.optim as optim

import ray
from ray import tune
from ray.tune.schedulers.test_scheduler import TestScheduler
from ray.tune.examples.mnist_pytorch import (train, test, get_data_loaders,
                                             ConvNet)
import ray.utils


# Change these values if you want the training to run quicker or slower.
EPOCH_SIZE = 512
TEST_SIZE = 256

# Training settings
parser = argparse.ArgumentParser(description="PyTorch MNIST Example")
parser.add_argument(
    "--use-gpu",
    action="store_true",
    default=False,
    help="enables CUDA training")
parser.add_argument(
    "--ray-address", type=str, help="The Redis address of the cluster.")
parser.add_argument(
    "--smoke-test", action="store_true", help="Finish quickly for testing")


# Below comments are for documentation purposes only.
# yapf: disable
# __trainable_example_begin__
class TrainMNIST(tune.Trainable):
    def setup(self, config):
        use_cuda = config.get("use_gpu") and torch.cuda.is_available()
        self.device = torch.device("cuda" if use_cuda else "cpu")
        self.train_loader, self.test_loader = get_data_loaders()
        self.model = ConvNet().to(self.device)
        self.optimizer = optim.SGD(
            self.model.parameters(),
            lr=config.get("lr", 0.01),
            momentum=config.get("momentum", 0.9))

    def step(self):
        tune.utils.wait_for_gpu(target_util=0.1)
        train(
            self.model, self.optimizer, self.train_loader, device=self.device)
        acc = test(self.model, self.test_loader, self.device)
        return {"mean_accuracy": acc}

    def save_checkpoint(self, checkpoint_dir):
        print("++++++++++++++++" + checkpoint_dir)
        checkpoint_path = os.path.join(checkpoint_dir, "model.pth")
        torch.save(self.model.state_dict(), checkpoint_path)
        return checkpoint_path

    def load_checkpoint(self, checkpoint_path):
        self.model.load_state_dict(torch.load(checkpoint_path))


# __trainable_example_end__
# yapf: enable



if __name__ == "__main__":
    args = parser.parse_args()
    ray.init(address=args.ray_address, num_cpus=6 if args.smoke_test else None)

    test_scheduler = TestScheduler()
    
    lr_list = list()
    momentum_list = list()
    random.seed(1)
    for i in range(0, 2):
        lr_list.append(random.uniform(0.001, 0.1))

    for i in range(0, 3):
        momentum_list.append(random.uniform(0.1, 0.9))

    analysis = tune.run(
        TrainMNIST,
        metric="mean_accuracy",
        mode="max",
        scheduler=test_scheduler,
        stop={
            # "mean_accuracy": 0.99,
            "training_iteration": 3 if args.smoke_test else 20,
        },
        resources_per_trial={
            "cpu": 1,
            "gpu": 1
        },
        num_samples=1 if args.smoke_test else 1,
        checkpoint_at_end=True,
        checkpoint_freq=3,
        config={
            "args": args,
            "lr": tune.grid_search(lr_list),
            "momentum": tune.grid_search(momentum_list),
        })

    print("Best config is:", analysis.get_best_config(metric="mean_accuracy", mode="max"))
```

thx

**Comment by ch1ck** [comment]
> Can you post your full script? I'm not quite sure why you're seeing this behavior. Pausing should release all resources of a given trial.

Hey @richardliaw, could you find anything wrong with my script?
In this example, I use grid_search to start 6 trials. Stop one of the trials and pause the rest of the trials in on_trial_result(). And the paused trials will not resume in choose_trial_to_run(). 
Then you may see the weird behavior.

**Comment by xwjiang2010** [comment]
I was able to reproduce the issue.
In my case, my cluster has 20CPUs running 6 trials, each taking one. 
```
caccc_00000 PAUSED
caccc_00001 PAUSED
caccc_00002 PAUSED
caccc_00003 TERMINATED
caccc_00004 PAUSED
caccc_00005 PAUSED
```
Although 5 of the trials are paused but available resource is only 15 (20 - 5).

It seems our logic does call `remove_placement_group` (API into Core) successfully. For example the following is the print statement at `_stop_trial`:
```
{'node:172.31.20.54': 1.0, 'CPU': 20.0, 'memory': 57821539531.0, 'object_store_memory': 24245708387.0, 'CPU_group_0_30ca138030a6ab86b0449c0c15c00edc': 1.0, 'CPU_group_45b6ff8dcb163478dcdc8a93c1f7cfdf': 1.0, 'CPU_group_30ca138030a6ab86b0449c0c15c00edc': 1.0, 'CPU_group_0_45b6ff8dcb163478dcdc8a93c1f7cfdf': 1.0, 'bundle_group_30ca138030a6ab86b0449c0c15c00edc': 1000.0, 'bundle_group_0_45b6ff8dcb163478dcdc8a93c1f7cfdf': 1000.0, 'bundle_group_45b6ff8dcb163478dcdc8a93c1f7cfdf': 1000.0, 'node:172.31.21.41': 1.0, 'bundle_group_0_30ca138030a6ab86b0449c0c15c00edc': 1000.0, 'node:172.31.27.235': 1.0, 'bundle_group_4b07d917f4cff11623c2687d91b1977c': 1000.0, 'bundle_group_0_4b07d917f4cff11623c2687d91b1977c': 1000.0, 'CPU_group_4b07d917f4cff11623c2687d91b1977c': 1.0, 'CPU_group_0_4b07d917f4cff11623c2687d91b1977c': 1.0, 'node:172.31.27.75': 1.0, 'CPU_group_4283ee5ef6f1a1a7b712abcbb5e885e4': 1.0, 'bundle_group_4283ee5ef6f1a1a7b712abcbb5e885e4': 1000.0, 'CPU_group_0_4283ee5ef6f1a1a7b712abcbb5e885e4': 1.0, 'bundle_group_0_4283ee5ef6f1a1a7b712abcbb5e885e4': 1000.0, 'CPU_group_0_63247fd1aea993ddd5c5c1a24623e96e': 1.0, 'CPU_group_63247fd1aea993ddd5c5c1a24623e96e': 1.0, 'bundle_group_63247fd1aea993ddd5c5c1a24623e96e': 1000.0, 'bundle_group_0_63247fd1aea993ddd5c5c1a24623e96e': 1000.0, 'node:172.31.22.165': 1.0}
```
And the placement to be removed for this trial is `4b07d917f4cff11623c2687d91b1977c`.
The next round at `_stop_trial` for the next trial to be paused:
```
{'memory': 57821539531.0, 'CPU_group_c686c4d56f4ea4a731acb7d5ad28c0f9': 1.0, 'object_store_memory': 24245708387.0, 'bundle_group_0_c686c4d56f4ea4a731acb7d5ad28c0f9': 1000.0, 'CPU_group_0_c686c4d56f4ea4a731acb7d5ad28c0f9': 1.0, 'bundle_group_c686c4d56f4ea4a731acb7d5ad28c0f9': 1000.0, 'CPU': 20.0, 'node:172.31.20.54': 1.0, 'bundle_group_45b6ff8dcb163478dcdc8a93c1f7cfdf': 1000.0, 'bundle_group_0_30ca138030a6ab86b0449c0c15c00edc': 1000.0, 'CPU_group_30ca138030a6ab86b0449c0c15c00edc': 1.0, 'bundle_group_0_45b6ff8dcb163478dcdc8a93c1f7cfdf': 1000.0, 'node:172.31.21.41': 1.0, 'CPU_group_45b6ff8dcb163478dcdc8a93c1f7cfdf': 1.0, 'CPU_group_0_30ca138030a6ab86b0449c0c15c00edc': 1.0, 'CPU_group_0_45b6ff8dcb163478dcdc8a93c1f7cfdf': 1.0, 'bundle_group_30ca138030a6ab86b0449c0c15c00edc': 1000.0, 'node:172.31.27.235': 1.0, 'bundle_group_4283ee5ef6f1a1a7b712abcbb5e885e4': 1000.0, 'node:172.31.27.75': 1.0, 'CPU_group_0_4283ee5ef6f1a1a7b712abcbb5e885e4': 1.0, 'CPU_group_4283ee5ef6f1a1a7b712abcbb5e885e4': 1.0, 'bundle_group_0_4283ee5ef6f1a1a7b712abcbb5e885e4': 1000.0, 'bundle_group_63247fd1aea993ddd5c5c1a24623e96e': 1000.0, 'node:172.31.22.165': 1.0, 'bundle_group_0_63247fd1aea993ddd5c5c1a24623e96e': 1000.0, 'CPU_group_0_63247fd1aea993ddd5c5c1a24623e96e': 1.0, 'CPU_group_63247fd1aea993ddd5c5c1a24623e96e': 1.0}
```

Compare the two, although `4b07d917f4cff11623c2687d91b1977c` is removed but `c686c4d56f4ea4a731acb7d5ad28c0f9` is added. This is strange to me, as I thought Tune's creation of PGs is already finished after all 6 of the trials runs. @krfricke Anything rings a bell?

**Comment by xwjiang2010** [comment]
cc @krfricke

**Comment by xwjiang2010** [comment]
@krfricke Agree with you. 

default: TUNE_REQUEST_PAUSED_TRIAL_RESOURCES = True if pbt else False

User can override this if they have more advanced scheduling algorithms.
